pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -348,10 +348,68 @@ func makeFirewallName(name string) string {
 	return fmt.Sprintf("k8s-fw-%s", name)
 }
 
+func (gce *GCECloud) getAddress(name, region string) (string, bool, error) {
+	address, err := gce.service.Addresses.Get(gce.projectID, region, name).Do()
+	if err == nil {
+		return address.Address, true, nil
+	}
+	if isHTTPErrorCode(err, http.StatusNotFound) {
+		return "", false, nil
+	}
+	return "", false, err
+}
+
+func ownsAddress(ip net.IP, addrs []*compute.Address) bool {
+	ipStr := ip.String()
+	for _, addr := range addrs {
+		if addr.Address == ipStr {
+			return true
+		}
+	}
+	return false
+}
+
 // EnsureTCPLoadBalancer is an implementation of TCPLoadBalancer.EnsureTCPLoadBalancer.
 // TODO(a-robinson): Don't just ignore specified IP addresses. Check if they're
 // owned by the project and available to be used, and use them if they are.
 func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, externalIP net.IP, ports []*api.ServicePort, hosts []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error) {
+	if externalIP == nil {
+		glog.V(2).Info("Checking if the external ip address already exists: %s", name)
+		address, exists, err := gce.getAddress(name, region)
+		if err != nil {
+			return nil, fmt.Errorf("error looking for gce address: %v", err)
+		}
+		if !exists {
+			// Note, though static addresses that _aren't_ in use cost money, ones that _are_ in use don't.
+			// However, quota is limited.  TODO: determine quota here
```

Comment:
note the new failure mode to handle - if the "upgrade" fails now it might
succeed later :)

On Fri, Oct 9, 2015 at 10:05 AM, Alex Robinson <notifications@github.com>
wrote:

> In pkg/cloudprovider/providers/gce/gce.go
> <https://github.com/kubernetes/kubernetes/pull/13335#discussion_r41653401>
> :
>
> >  // EnsureTCPLoadBalancer is an implementation of TCPLoadBalancer.EnsureTCPLoadBalancer.
> >  // TODO(a-robinson): Don't just ignore specified IP addresses. Check if they're
> >  // owned by the project and available to be used, and use them if they are.
> >  func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, externalIP net.IP, ports []*api.ServicePort, hosts []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error) {
> > +	if externalIP == nil {
> > +		glog.V(2).Info("Checking if the external ip address already exists: %s", name)
> > +		address, exists, err := gce.getAddress(name, region)
> > +		if err != nil {
> > +			return nil, fmt.Errorf("error looking for gce address: %v", err)
> > +		}
> > +		if !exists {
> > +			// Note, though static addresses that _aren't_ in use cost money, ones that _are_ in use don't.
> > +			// However, quota is limited.  TODO: determine quota here
>
> We want this in 1.1. It came up when we were discussing the pain of
> upgrading people's existing clusters that may not have enough static IP
> quota to satisfy their existing services' needs.
>
> I'll work it into #15286
> <https://github.com/kubernetes/kubernetes/pull/15286>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13335/files#r41653401>.
>